### ISSUE TITLE
[E::bam_hdr_write] Header too long for BAM format; [main_samview] failed to write the SAM header

### ISSUE THREAD

**Issue by rcmueller** [samtools/samtools#1613]
#### Are you using the latest version of [samtools](https://github.com/samtools/samtools/releases/latest) and [HTSlib](https://github.com/samtools/htslib/releases/latest)? If not, please specify.
(run `samtools --version`)
`samtools 1.15` (but the same issue with 1.14)
`Using htslib 1.15`

 #### Please describe your environment.
- OS (run `uname -sr` on Linux/Mac OS or ` wmic os get Caption, Version` on Windows)
`Linux 5.16.9-200.fc35.x86_64`
- machine architecture (run `uname -m` on Linux/Mac OS or ` wmic os get OSArchitecture` on Windows)
`x86_64`
- compiler (run `gcc --version` or `clang --version`)
`gcc (GCC) 11.2.1 20220127 (Red Hat 11.2.1-9)`

#### Please specify the steps taken to generate the issue, the command you are running and the relevant output.

Dear community,

I would like to convert a large sam file created by bowtie2 to bam (and subsequently sort/index ... downstream process):
`samtools view file.sam -o file.bam`

However, I get the following error:
`[E::bam_hdr_write] Header too long for BAM format`
`[main_samview] failed to write the SAM header`

I guess, this error refers to the overall size of the header (as opposed to line length).
Smaller sam files (~ 20 Mio header lines) produced the same way can be converted/processed.

File size: `33 Gb`
Header lines (`samtools view -H file.sam | wc -l`): `45560215`
Header `samtools view -H file.sam | tail -n 5` (paths in PG truncated):
`@SQ  SN:A00266:279:HMKMLDSXX:3:2668:25003:9314|44504 LN:151`
`@SQ     SN:A00266:279:HMKMLDSXX:3:2670:6804:36448|44504 LN:151`
`@SQ     SN:A00266:279:HMKMLDSXX:3:2677:13413:28291|44504        LN:151`
`@PG     ID:bowtie2      PN:bowtie2      VN:2.4.5        CL:"bowtie2-2.4.5-linux-x86_64/bowtie2-align-l --wrapper basic-0 -p 80 --local -x REFERENCE --passthrough -1 R1.fastq.gz -2 R2.fastq.gz"`
`@PG     ID:samtools     PN:samtools     PP:bowtie2      VN:1.15 CL:samtools view -H file.sam`

I could not find a hint on the web, in issues or relevant forums, and neither could I find header size limits in the specifications.
So I was hoping you could help. I am happy to provide further details.

Best wishes,
Ralf

**Comment by whitwham** [comment]
The header size limit is on page 15 of the SAM spec:

`l_text, Length of the header text, including any NUL padding, uint32 t, < 2^31`

Though I am not surprised you did not find it.  This is a BAM limitation due to its binary nature rather than a SAM one.

**Comment by daviesrob** [comment]
With `samtools 1.15` you can work around this problem by saving your files in `sam.gz` format.  `samtools` should mostly work with these in the same way as for BAM files, but support in other tools may vary.

The references in your file look a lot like individual sequencing reads.  Is that as you expect?

**Comment by jkbonfield** [comment]
I think @daviesrob hit the nail on the head here.  Your file looks like a naive "fastq to SAM" conversion tool, probably designed for a different purpose.  I've seen such atrocities in blogs before, but when people say converting fastq to SAM they normally mean (but sadly not saying) sequence alignment with e.g. bwa, minimap, bowtie, novoalign, etc.

**Comment by rcmueller** [comment]
Thank you for your swift responses and clarifications!
In fact, I saw that limit (<2^31), but for some strange reason, I did not make the link to 40 Mio lines not working and 20 Mio where it worked. @jmarshall That's perfectly right, the lines are all around 60 characters long, and the result of bowtie2 mapping pre-processed raw reads to predicted and annotated genes based on assembled contigs.
Thanks for the idea @jmarshall, the workaround might actually work.

Best,
Ralf
